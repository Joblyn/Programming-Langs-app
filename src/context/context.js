import { createContext } from "react";

export const RouteContext = createContext(null);

export const NameContext = createContext(null);