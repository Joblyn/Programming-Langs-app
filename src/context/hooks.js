export const useAppContext = (value = null) => {
  let route = value;
  return route;
};