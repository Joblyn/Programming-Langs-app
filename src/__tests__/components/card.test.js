import React from "react";
import { Router } from "react-router-dom";
import { render, cleanup, fireEvent } from "@testing-library/react";
import Languages from "../../data/languages";
import { AppContext } from "../../context/context";
import CustomCard from "../../components/card/card";
import "@testing-library/jest-dom";
import { createMemoryHistory } from "history";

// mock react-router-dom
jest.mock("react-router-dom");

// mock useNavigate from React
const mockedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedNavigate,
}));

// mock useState setState from React
const setRoute = jest.fn();
const useStateMock = jest.spyOn(React, "useState");
useStateMock.mockImplementation((initialState) => [initialState, setRoute]);

// testing card component
describe("<CustomCard />", () => {
  afterEach(cleanup);

  // UI test
  it("renders the component correctly", () => {
    // initial memoryHistory
    const history = createMemoryHistory();
    const { container, getByText, getByAltText } = render(
      <Router location={history.location}>
        <AppContext.Provider value={{ setRoute }}>
          <CustomCard
            title={Languages[0].title}
            image={Languages[0].image}
            alt={Languages[0].alt}
          />
        </AppContext.Provider>
      </Router>
    );

    expect(getByText(Languages[0].title)).toBeTruthy();
    expect(getByAltText(Languages[0].alt)).toBeTruthy();
    expect(container.firstChild.matches(".custom-card")).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });

  // triggers setRoute upon click
  it("sets the route upon click and changes pathname to name", async () => {
     // initialise memoryHistory
     const history = createMemoryHistory();
    const { container, getByText } = render(
      <Router location={history.location} navigator={history}>
        <AppContext.Provider value={{ setRoute }}>
          <CustomCard
            title={Languages[0].title}
            image={Languages[0].image}
            alt={Languages[0].alt}
            href={Languages[0].href}
          />
        </AppContext.Provider>
      </Router>
    );

    let button = container.getElementsByTagName("button")[0];
    await fireEvent.click(button);

    // check that the route is set
    expect(setRoute).toHaveBeenCalledWith(Languages[0].href);
    expect(mockedNavigate).toHaveBeenCalledWith("/assessment/name");
  });
});
