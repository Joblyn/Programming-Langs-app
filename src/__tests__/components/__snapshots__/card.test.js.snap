// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomCard /> renders the component correctly 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root w-80 custom-card css-1ri6ub7-MuiPaper-root-MuiCard-root"
>
  <button
    class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
    tabindex="0"
    type="button"
  >
    <img
      alt="Shell Scripting"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
      height="140"
      src="shell_scripting.jpeg"
      style="height: 200px;"
      width="345"
    />
    <div
      class="MuiCardContent-root bg-white css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiTypography-root MuiTypography-p MuiTypography-gutterBottom text-xl css-anmq6p-MuiTypography-root"
      >
        Shell Scripting
      </div>
    </div>
    <span
      class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
    />
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
